server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/DKIMSign.java

```diff
@@ -0,0 +1,182 @@
+ *   ycY0kjsdxGKDudWcsATllFzXDCQF6DTNIWECQEA52ePwTjKrVnLTfCLEG4OgHKvl
+ *   Zud4amthwDyJWoMEH2ChNB2je1N4JLrABOE+hk+OuoKnKAKEjWd8f3Jg/rkCQHj8
+ *   mQmogHqYWikgP/FSZl518jV48Tao3iXbqvU9Mo2T6yzYNCCqIoDLFWseNVnCTZ0Q
+ *   b+IfiEf1UeZVV5o4J+ECQDatNnS3V9qYUKjj/krNRD/U0+7eh8S2ylLqD3RlSn9K
+ *   tYGRMgAtUXtiOEizBH6bd/orzI9V9sw8yBz+ZqIH25Q=
+ *   -----END RSA PRIVATE KEY-----
+ *   &lt;/privateKey&gt;
+ * &lt;/mailet&gt;
+ * </code></pre>
+ * 
+ * By default the mailet assume that Javamail will convert LF to CRLF when sending 
+ * so will compute the hash using converted newlines. If you don't want this 
+ * behaviout then set forceCRLF attribute to false. 
+ */
+public class DKIMSign extends GenericMailet {
+
+    private String signatureTemplate;
+    private PrivateKey privateKey;
+    private boolean forceCRLF;
+
+	/**
+	 * @return the signatureTemplate
+	 */
+	protected String getSignatureTemplate() {
+		return signatureTemplate;
+	}
+
+	/**
+	 * @return the privateKey
+	 */
+	protected PrivateKey getPrivateKey() {
+		return privateKey;
+	}
+
+    public void init() throws MessagingException {
+        signatureTemplate = getInitParameter("signatureTemplate");
+        String privateKeyString = getInitParameter("privateKey");
+        String privateKeyPassword = getInitParameter("privateKeyPassword", null);
+        forceCRLF = getInitParameter("forceCRLF", true);
+        try {
+            PKCS8Key pkcs8 = new PKCS8Key(new ByteArrayInputStream(
+                    privateKeyString.getBytes()),
+                    privateKeyPassword != null ? privateKeyPassword
+                            .toCharArray() : null);
+            privateKey = pkcs8.getPrivateKey();
+            // privateKey = DKIMSigner.getPrivateKey(privateKeyString);
+        } catch (NoSuchAlgorithmException e) {
+            throw new MessagingException("Unknown private key algorythm: "
+                    + e.getMessage(), e);
+        } catch (InvalidKeySpecException e) {
+            throw new MessagingException(
+                    "PrivateKey should be in base64 encoded PKCS8 (der) format: "
+                            + e.getMessage(), e);
+        } catch (GeneralSecurityException e) {
+            throw new MessagingException("General security exception: "
+                    + e.getMessage(), e);
+        }
+    }
+
+    public void service(Mail mail) throws MessagingException {
+        DKIMSigner signer = new DKIMSigner(getSignatureTemplate(), getPrivateKey());
+        SignatureRecord signRecord = signer
+                .newSignatureRecordTemplate(getSignatureTemplate());
+        try {
+            BodyHasher bhj = signer.newBodyHasher(signRecord);
+            MimeMessage message = mail.getMessage();
+            Headers headers = new MimeMessageHeaders(message);
+            try {
+            	OutputStream os = new HeaderSkippingOutputStream(bhj.getOutputStream());
+            	if (forceCRLF) os = new CRLFOutputStream(os);
```

Comment:
Moving is one issue, refactoring is an other. Do we want to take time doing this?

(It is an external contribution, we can ask Shrumit to do it...)